File: ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 4
end

class Squares
  def initialize(n)
    @range = 0..n
  end

  def sum_of_squares
    @range.reduce {|sum, x| sum + x**2}
  end

  def square_of_sum
    @range.reduce(:+) ** 2
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: r/difference-of-squares/difference-of-squares.R
square_of_sum <- function(n) {
  Reduce(sum, 1:n) ** 2
}

sum_of_squares <- function(n) {
  Reduce(function(acc, x) acc + x ** 2, 1:n)
}

difference_of_squares <- function(n) {
  square_of_sum(n) - sum_of_squares(n)
}
